coordinator/history/history.go

```diff
@@ -0,0 +1,231 @@
+// Copyright 2024 Edgeless Systems GmbH
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package history
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+
+	"github.com/spf13/afero"
+)
+
+const (
+	hashSize = 32 // byte, History.hashFun().Size()
+	histPath = "/mnt/state/history"
+)
+
+// History is the history of the Coordinator.
+type History struct {
+	store      store
+	hashFun    func() hash.Hash
+	signingKey *ecdsa.PrivateKey
+}
+
+// New creates a new History with the given signing key.
+func New() (*History, error) {
+	osFS := afero.NewOsFs()
+	if err := osFS.MkdirAll(histPath, 0o755); err != nil {
+		return nil, fmt.Errorf("creating history directory: %w", err)
+	}
+	h := &History{
+		store:   newPVStore(&afero.Afero{Fs: afero.NewBasePathFs(osFS, histPath)}),
+		hashFun: sha256.New,
+	}
+	if hashSize != h.hashFun().Size() {
+		return nil, errors.New("mismatch between hashSize and hash function size")
+	}
+	return h, nil
+}
+
+// ConfigureSigningKey sets the signing key for validation and signing of the protected history parts.
+func (h *History) ConfigureSigningKey(signingKey *ecdsa.PrivateKey) {
+	h.signingKey = signingKey
+}
+
+// GetManifest returns the manifest for the given hash.
+func (h *History) GetManifest(hash [hashSize]byte) ([]byte, error) {
+	return h.getContentaddressed("manifests/%s", hash)
+}
+
+// SetManifest sets the manifest and returns its hash.
+func (h *History) SetManifest(manifest []byte) ([hashSize]byte, error) {
+	return h.setContentaddressed("manifests/%s", manifest)
+}
+
+// GetPolicy returns the policy for the given hash.
+func (h *History) GetPolicy(hash [hashSize]byte) ([]byte, error) {
+	return h.getContentaddressed("policies/%s", hash)
+}
+
+// SetPolicy sets the policy and returns its hash.
+func (h *History) SetPolicy(policy []byte) ([hashSize]byte, error) {
+	return h.setContentaddressed("policies/%s", policy)
+}
+
+// GetTransition returns the transition for the given hash.
+func (h *History) GetTransition(hash [hashSize]byte) (*Transition, error) {
+	transitionBytes, err := h.getContentaddressed("transitions/%s", hash)
+	if err != nil {
+		return nil, err
+	}
+	var transition Transition
+	if err := transition.unmarshalBinary(transitionBytes); err != nil {
+		return nil, fmt.Errorf("unmarshaling transition: %w", err)
+	}
+	return &transition, nil
+}
+
+// SetTransition sets the transition and returns its hash.
+func (h *History) SetTransition(transition *Transition) ([hashSize]byte, error) {
+	return h.setContentaddressed("transitions/%s", transition.marshalBinary())
+}
+
+// GetLatest returns the verified transition for the given hash.
+func (h *History) GetLatest() (*LatestTransition, error) {
```

Comment:
While implementing I realized that I need to be able to tell the difference between generic errors and "no latest set".

Reply:
The issue is that I'd like to update my local state from persistency before handling a SetManifest, when I don't yet have something to CompareAndSet. But I can work around that by ignoring errors when my local state suggests emptiness - I will be stopped at the very latest when my update conflicts. So this would be a good extension, but not necessary right now.

Reply:
use `errors.Is(err, os.ErrNotExist)` ?

Reply:
I can do that for now, but it's somewhat counter to the idea of having pluggable KV stores :)

Reply:
True.